src/openms/include/OpenMS/ANALYSIS/TOPDOWN/SpectrumDeconvolution.h

```diff
@@ -0,0 +1,159 @@
+    //Peaks with log mz transformation
+    typedef FLASHDeconvHelperStructs::LogMzPeak LogMzPeak;
+
+    /// default constructor
+    SpectrumDeconvolution(MSSpectrum &spec, Parameter &param);
+
+    /// default destructor
+    ~SpectrumDeconvolution();
+
+    bool empty();
+
+    //Main function
+    std::vector<PeakGroup> &getPeakGroupsFromSpectrum(std::vector<std::vector<Size>> &prevMassBinVector,
+                                                      std::vector<double> &prevMinBinLogMassVector,
+                                                      FLASHDeconvHelperStructs::PrecalculatedAveragine &avg,
+                                                      unsigned int msLevel);
+
+    //Stores log mz peaks
+    std::vector<LogMzPeak> logMzPeaks;
+
+  protected:
+    MSSpectrum &spec;
+    Parameter &param;
+
+    //Bins for mass and mz. Bin size is deteremine by Parameter.tolerance
+    //massBinsForThisSpectrum stores the selected bins only for this spectrum
+    boost::dynamic_bitset<> massBinsForThisSpectrum;
+    //massBins stores the selected bins for this spectrum + overlapped spectrum (previous a few spectra).
+    boost::dynamic_bitset<> massBins;
+    //mzBins stores the binned log mz peaks
+    boost::dynamic_bitset<> mzBins;
+    //peakGroups stores the decovnoluted mass peak groups
+    std::vector<PeakGroup> peakGroups;
+
+    //This stores the "universal pattern"
+    double *filter{};
+    //This stores the patterns for harmonic reduction
+    double **harmonicFilter{};
+
+    //This stores the "universal pattern" in binned dimenstion
+    long *binOffsets{};
+    //This stores the patterns for harmonic reduction in binned dimenstion
+    long **hBinOffsets{};
+
+    //static fucntion that converts bin to value
+    static double getBinValue(Size bin, double minV, double binWidth);
+
+    //static function that converts value to bin
+    static Size getBinNumber(double v, double minV, double binWidth);
+
+    static void pringMasses(boost::dynamic_bitset<> &massBins, double &minMass, double binWidth);
+
+    //generate log mz peaks from the input spectrum
+    void updateLogMzPeaks(double chargeMass);
+
+    //generate mz bins from log mz peaks
+    void updateMzBins(double &mzBinMinValue, Size &binNumber, double binWidth,
+                      float *mzBinIntensities);
+
+    //this function takes the previous deconvolution results (from ovelapped spectra) for sensitive deconvolution of the current spectrum
+    void unionPrevMassBins(double &massBinMinValue,
+                           std::vector<std::vector<Size>> &prevMassBinVector,
+                           std::vector<double> &prevMassBinMinValue,
+                           UInt msLevel);
+
+    //Update mass bins from mz bins and universal pattern. It select candidate mass bins using the pattern, eliminate possible harmonic masses
+    std::vector<Byte *> updateMassBins(double &massBinMinValue,
```

Comment:
can we get rid of the Byte* ?

Reply:
Is it OK to use vector<vector<Byte>> then? Or Byte is the problem?

Reply:
you could use Matrix<Byte> in OpenMS. This might even be faster because all bytes will occupy consecutive memory addresses.

Reply:
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/DATASTRUCTURES/Matrix.h